_use_cases/encryption-at-rest.markdown

```diff
@@ -0,0 +1,105 @@
+the encryption key so won't be able to read the data, including the Kafka records.
+
+However, there are shortcomings.  The encrypted storage volume must be mounted to the computers running the Kafka broker. To do
+this the  and the encryption keys applied at the operating system level.  Anybody with shell level access to the hosts  is likely
+to be able to read the data, including the Kafka confidential records.
+
+Storage volume encryption doesn't really solve the problem.
+
+### Can't the applications encrypt/decrypt the data?
+
+It is possible for producing applications to encrypt data before sending it to Kafka, and for consuming applications to decrypt it
+again.  With this approach the brokers never possess the records in the clear and as they don't have encryption keys, they cannot
+decrypt it.
+
+So, this approach does offer a solution to the problem but there are disadvantages.
+
+1. Kafka Client libraries don't support encryption themselves.
+1. Kafka Client ecosystem is polyglot - it is common to Kafka applications written in many languages (Java, Go, Rust...)
+   deployed within a _single organisation_.  Any solution needs to work across all the target languages.
+1. There are some encryption libraries for some languages that can be used with the Kafka Clients but these available
+   across all languages.
+
+Organisations could write their own encryption code but this is a burden to the application teams. The need for
+_interoperability_ between different language encryption implementations - this increases the overheads.  There is also
+the problem of key distribution to consider.  Some part of the system needs to push the correct encryption keys out to
+the applications and manage tasks such as key rotation.  Finally, cryptography is a specialist area and the consequences
+of a design flaw or bug are significant (confidentiality breach).
+
+Having the applications encrypt/decrypt data themselves, whilst technically feasible, is not really a tenable solution
+at the scale required for most enterprises.
+
+# Kroxylicious Topic Encryption
+
+The Kroxylicious Topic Encryption feature offers a solution to the problem.  The proxy takes the responsibility
+to encrypt and decrypt the messages. In this way, the Kafka Brokers never see the plain text content of
+the message thus ensuring confidentiality.  Encryption is introduced into the system without requiring changes to
+either applications or the Kafka Cluster.
+
+Within Kroxylicious, the job of encryption and decryption is delegated to a Kroxylicious Filter. The filter intercepts
+produce requests from producing applications and encrypts the content, before the produce request is forwarded to the
+brokers.  For consuming, the filter intercepts fetch responses and decrypts the contents before the
+fetch response is forwarded to the applications.
+
+As the solution is proxy-based, it works regardless of the language the applications is written in.
```

Comment:
The other advantage to the proxy is it doesn't matter where the application is deployed either. It can be deployed in a different security context to the proxy and to the Kafka cluster. Thus we are adding more flexibility.